### PR TITLE
FOGL-2098 bulk config items end point added

### PR DESCRIPTION
**Pull request by ashish-jabble**
- [x] PUT `/foglamp/category/{category_name}` new endpoint added for bulk config item save in single call
- [x] Add Configuration API Unit tests 
- [x] Add Configuration Manager Unit tests 
- [x] JIRA added for GUI to implement https://scaledb.atlassian.net/browse/FOGL-2106

### CODE REVIEW

File: python/foglamp/common/configuration_manager.py [hunk 1]
```diff
@@ -426,6 +426,71 @@ def get_entry_val(k):
...
+
+        Returns:
+            None
+        """
+
+        try:
+            payload = {"updates": []}
+            audit_details = {'category': category_name, 'items': {}}
+            for item_name, new_val in config_item_list.items():
+                if not isinstance(new_val, str):
```
Comment (praveen-garg) [inline python/foglamp/common/configuration_manager.py]: Allow dict for JSON

Comment (praveen-garg) [reply python/foglamp/common/configuration_manager.py]: actually, _validate_type_value is already there hence you can skip this check at all.

Comment (ashish-jabble) [reply python/foglamp/common/configuration_manager.py]: > actually, _validate_type_value is already there hence you can skip this check at all.

Nope, we should restrict before if the expected value is not in string format. JSON is an exceptional case.

File: python/foglamp/common/configuration_manager.py [hunk 2]
```diff
@@ -426,6 +426,71 @@ def get_entry_val(k):
...
+                    raise TypeError('new value should be of string type')
+
+                cat_item = await self._read_item_val(category_name, item_name)
```
Comment (praveen-garg) [inline python/foglamp/common/configuration_manager.py]: are we querying database for each item? We should read once for given category. and then use that for iteration

File: python/foglamp/common/configuration_manager.py (hunk 1 above)
Comment (praveen-garg) [reply python/foglamp/common/configuration_manager.py]: I'm not sure what is the value of checking twice. what is the use of _validate_type_value

Comment (ashish-jabble) [reply python/foglamp/common/configuration_manager.py]: **_validate_type_value** converts string to type as given. If value not "string" then it raises an Exception

Comment (praveen-garg) [reply python/foglamp/common/configuration_manager.py]: Yeah, that is good enough!

File: python/foglamp/services/core/routes.py [hunk 3]
```diff
@@ -64,6 +64,7 @@ def setup(app):
     app.router.add_route('GET', '/foglamp/category/{category_name}/children', api_configuration.get_child_category)
     app.router.add_route('DELETE', '/foglamp/category/{category_name}/children/{child_category}', api_configuration.delete_child_category)
     app.router.add_route('DELETE', '/foglamp/category/{category_name}/parent', api_configuration.delete_parent_category)
+    app.router.add_route('PUT', '/foglamp/category/{category_name}/value', api_configuration.update_configuration_item_bulk)
```
Comment (praveen-garg) [inline python/foglamp/services/core/routes.py]: I think, We should drop `/value` from this url.

File: python/foglamp/common/configuration_manager.py (hunk 2 above)
Comment (ashish-jabble) [reply python/foglamp/common/configuration_manager.py]: Done

File: python/foglamp/services/core/routes.py (hunk 3 above)
Comment (MarkRiddoch) [reply python/foglamp/services/core/routes.py]: Agreed

Comment (ashish-jabble) [reply python/foglamp/services/core/routes.py]: Done

File: tests/unit/python/foglamp/common/test_configuration_manager.py [hunk 4]
```diff
@@ -2192,3 +2192,78 @@ def mock_coro():
...
+            return return_value
+
+        cat_info = {'enableHttp': {'default': 'true', 'description': 'Enable HTTP', 'type': 'boolean', 'value': 'true'}}
+        config_item_list = {"enableHttp": "true"}
+        storage_client_mock = MagicMock(spec=StorageClientAsync)
+        c_mgr = ConfigurationManager(storage_client_mock)
+        with patch.object(c_mgr, 'get_category_all_items', return_value=async_mock(cat_info)) as patch_get_all_items:
+            result = await c_mgr.update_configuration_item_bulk(category_name, config_item_list)
+            assert result is None
+        patch_get_all_items.assert_called_once_with(category_name)
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/common/test_configuration_manager.py]: should not we add `patch_update` not called (and may be also not the patch_callback, patch_audit, patch_read_val)

File: tests/unit/python/foglamp/common/test_configuration_manager.py [hunk 5]
```diff
@@ -2192,3 +2192,78 @@ def mock_coro():
...
+        storage_client_mock = MagicMock(spec=StorageClientAsync)
+        c_mgr = ConfigurationManager(storage_client_mock)
+        with patch.object(c_mgr, 'get_category_all_items', return_value=async_mock(cat_info)) as patch_get_all_items:
+            with patch.object(c_mgr._storage, 'update_tbl', return_value=async_mock(update_result)) as patch_update:
+                with patch.object(c_mgr, '_read_category_val', return_value=async_mock(read_val)) as patch_read_val:
+                    with patch.object(AuditLogger, '__init__', return_value=None):
+                        with patch.object(AuditLogger, 'information', return_value=async_mock(None)) as patch_audit:
+                            with patch.object(ConfigurationManager, '_run_callbacks', return_value=async_mock(None)) \
+                                    as patch_callback:
+                                await c_mgr.update_configuration_item_bulk(category_name, config_item_list)
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/common/test_configuration_manager.py]: should also test that cache updated

File: tests/unit/python/foglamp/services/core/api/test_configuration.py [hunk 6]
```diff
@@ -641,3 +641,47 @@ def async_mock():
...
+        storage_client_mock = MagicMock(spec=StorageClientAsync)
+        c_mgr = ConfigurationManager(storage_client_mock)
+        with patch.object(connect, 'get_storage_async', return_value=storage_client_mock):
+            with patch.object(c_mgr, 'update_configuration_item_bulk', side_effect=exception_name) as patch_update_bulk:
+                resp = await client.put('/foglamp/category/{}'.format(category_name), data=json.dumps(payload))
+                assert code == resp.status
+                assert resp.reason is None
+            patch_update_bulk.assert_called_once_with(category_name, payload)
+
+    async def test_update_bulk_config(self, client, category_name='rest_api'):
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/core/api/test_configuration.py]: we should also add a test for encoded category name like `S #1` => `S%20%231`

File: tests/unit/python/foglamp/services/core/api/test_configuration.py [hunk 7]
```diff
@@ -641,3 +641,50 @@ def async_mock():
...
+        c_mgr = ConfigurationManager(storage_client_mock)
+        with patch.object(connect, 'get_storage_async', return_value=storage_client_mock):
+            with patch.object(c_mgr, 'update_configuration_item_bulk', side_effect=exception_name) as patch_update_bulk:
+                resp = await client.put('/foglamp/category/{}'.format(category_name), data=json.dumps(payload))
+                assert code == resp.status
+                assert resp.reason is None
+            patch_update_bulk.assert_called_once_with(category_name, payload)
+
+    @pytest.mark.parametrize("category_name", [
+        "rest_api", "Rest $API"
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/core/api/test_configuration.py]: @pytest.mark.parametrize("category_name", [
        "rest_api", "Rest $API"

    @pytest.mark.parametrize("encoded_category_name", "category_name", [
        "rest_api", "Rest%20%24API", ["rest_api", "Rest $API"])

File: tests/unit/python/foglamp/services/core/api/test_configuration.py [hunk 8]
```diff
@@ -641,3 +641,50 @@ def async_mock():
...
+                  'enableHttp': {'description': 'Enable HTTP (disable to use HTTPS)', 'displayName': 'Enable HTTP', 'value': 'true', 'default': 'true', 'order': '1', 'type': 'boolean'},
+                  'httpPort': {'description': 'Port to accept HTTP connections on', 'displayName': 'HTTP Port', 'value': '8082', 'default': '8081', 'order': '2', 'type': 'integer'}}
+
+        payload = {"http_port": "8082", "authentication": "mandatory"}
+        storage_client_mock = MagicMock(spec=StorageClientAsync)
+        c_mgr = ConfigurationManager(storage_client_mock)
+        with patch.object(connect, 'get_storage_async', return_value=storage_client_mock):
+            with patch.object(c_mgr, 'update_configuration_item_bulk', return_value=async_mock(response)) as patch_update_bulk:
+                with patch.object(c_mgr, 'get_category_all_items', return_value=async_mock(result)) as patch_get_all_items:
+                    resp = await client.put('/foglamp/category/{}'.format(category_name), data=json.dumps(payload))
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/core/api/test_configuration.py]: encoded_category_name

File: tests/unit/python/foglamp/services/core/api/test_configuration.py (hunk 7 above)
Comment (praveen-garg) [reply tests/unit/python/foglamp/services/core/api/test_configuration.py]: Okay, This is not needed. 
> aiohttp internally performs URL canonization before sending request.

**Pull request merged**